napari/_qt/threading.py

```diff
@@ -0,0 +1,688 @@
+        - *finished*: emitted when the work is finished
+        - *returned*: emitted with return value
+        - *errored*: emitted with error object on Exception
+
+    If the decorated function is a generator, the (default) returned worker
+    will also provide these signals:
+
+        - *yielded*: emitted with yielded values (if generator used)
+        - *paused*: emitted when a running job has successfully paused
+        - *resumed*: emitted when a paused job has successfully resumed
+        - *aborted*: emitted when a running job is successfully aborted
+
+    And these methods:
+
+        - *quit*: ask the thread to quit
+        - *toggle_paused*: toggle the running state of the thread.
+        - *send*: send a value into the generator.  (This requires that your
+          decorator function uses the ``value = yield`` syntax)
+
+    If you call your function with ``func(_start_thread=False)`` (the default)
+
+        - *start*: it will also have a ``.start()`` method that can be used to
+        start execution of the function in another thread.  (useful if you need
+        to connect callbacks to signals prior to execution)
+
+    Parameters
+    ----------
+    func : callable
+        Function to call in another thread.  For communication between threads
+        may be a generator function.
+
+    Returns
+    -------
+    callable
+        function that creates a worker, puts it in a new thread and returns
+        the worker instance.
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        @thread_worker
+        def long_function(start, end):
+            # do setup here.
+            yield # Critical to yield before computation when using generator
+
+            # do work, periodically yielding
+            i = start
+            while i <= end:
+                time.sleep(0.1)
+                yield i
+
+            # do teardown
+            return 'anything'
+
+        # call the function to start running in another thread.
+        worker = long_function(_start_thread=False)
+        # connect signals if desired
+        worker.start()
+    """
+
+    # this extra level of nesting (and the "optional" function in the sig)
+    # allows this decorator to be used with OR without arguments:
+    # @thread_worker
+    # or
+    # @thread_worker(start_thread=True)
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # decorator kwargs can be overridden at call time by using the
+            # underscore-prefixed version of the kwarg.
+            kwargs['_start_thread'] = kwargs.get('_start_thread', start_thread)
+            kwargs['_connect'] = kwargs.get('_connect', connect)
+            kwargs['_worker_class'] = kwargs.get('_worker_class', worker_class)
+            return create_worker(func, *args, **kwargs,)
+
+        return wrapper
+
+    if function is None:
+        return decorator
+    return decorator(function)
```

Comment:
Ooooooh, right. I use `toolz.curry` for this, I think it ends up way prettier:

```python
@tz.curry
def thread_worker(function, start_thread=False, connect=None, worker_class=None):
    @functools.wraps
    def worker_function(*args, **kwargs):
        kwargs['_start_thread'] = kwargs.get('_start_thread', start_thread)
        kwargs['_connect'] = kwargs.get('_connect', connect)
        kwargs['_worker_class'] = kwargs.get('_worker_class', worker_class)
        return create_worker(func, *args, **kwargs)
    return worker_function
```

Reply:
Omg 😍